server/src/main/java/io/crate/replication/logical/action/ReplayChangesAction.java

```diff
@@ -101,20 +112,63 @@ protected void shardOperationOnPrimary(Request request,
+                primary.advanceMaxSeqNoOfUpdatesOrDeletes(request.maxSeqNoOfUpdatesOrDeletes());
+            }
+            try {
+                return primary.applyTranslogOperation(opWithPrimary, Engine.Operation.Origin.PRIMARY);
+            } catch(IOException e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        private void waitMappingUpdateInClusterState(Translog.Operation op, IndexShard primary, Request request, Exception failure, Consumer<Engine.Result> collectResult) {
+            final var indexName = request.shardId().getIndexName();
+            Metadata metadata = clusterService.state().metadata();
+            var currentMappingVersion = metadata.index(indexName).getMappingVersion();
+            ClusterStateObserver clusterStateObserver = new ClusterStateObserver(clusterService, LOGGER);
+            clusterStateObserver.waitForNextChange(
+                new ClusterStateObserver.Listener() {
+                    @Override
+                    public void onNewClusterState(ClusterState state) {
+                        var result = applyTransLogOperation(op, primary, request);
+                        collectResult.accept(result);
+                    }
+
+                    @Override
+                    public void onClusterServiceClose() {}
+
+                    @Override
+                    public void onTimeout(TimeValue timeout) {
+                        LOGGER.warn(failure);
+                    }
+
+                }, cs -> isIndexMetaDataOnClusterStateUpdated(cs, request.shardId().getIndexName(), currentMappingVersion)
+            );
+        }
+
+        private static boolean isIndexMetaDataOnClusterStateUpdated(ClusterState cs, String indexName, long mappingVersion) {
+            IndexMetadata indexMetadata = cs.metadata().index(indexName);
+            if (indexMetadata == null) {
+                return false;
+            } else {
+                var newMappingVersion = indexMetadata.getMappingVersion();
+                return newMappingVersion > mappingVersion;
+            }
+        }
+
         public Translog.Location performOnPrimary(Request request, IndexShard primary) throws Exception {
             Translog.Location location = null;
-
-            for (var translogOp : request.changes()) {
-                final var op = withPrimaryTerm(translogOp, primary.getOperationPrimaryTerm());
-                if (primary.getMaxSeqNoOfUpdatesOrDeletes() < request.maxSeqNoOfUpdatesOrDeletes()) {
-                    primary.advanceMaxSeqNoOfUpdatesOrDeletes(request.maxSeqNoOfUpdatesOrDeletes());
-                }
-                var result = primary.applyTranslogOperation(op, Engine.Operation.Origin.PRIMARY);
-                if (result.getResultType() == Engine.Result.Type.MAPPING_UPDATE_REQUIRED) {
-                    throw new RuntimeException("Mapping update required");
+            var result = new AtomicReference<Engine.Result>();
```

Comment:
I think technically every container would be enough here, as this is not multithreaded.

Reply:
Is there a better way to wait until the mappings are updated ?